clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -66,34 +82,75 @@ private URI translateUri(URI uri) throws java.net.URISyntaxException {
 
     @Override
     public void initialize(URI name, Configuration conf) throws IOException {
-        super.initialize(name, conf);
+	super.initialize(name, conf);
 	this.conf = conf;
-        LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ initialize: {} $$$$$$$$$$$$$$$$$$$$$$$$$$$$", name);
-
-        String host = name.getHost();
-        if (host == null) {
-            throw new IOException("Invalid repository specified");
-        }
-        setConf(conf);
-        this.uri = name;
-
-        // setup lakeFS api client
-        String endpoint = conf.get(FS_LAKEFS_ENDPOINT, "http://localhost:8000/api/v1");
-        String accessKey = conf.get(FS_LAKEFS_ACCESS_KEY);
-        if (accessKey == null) {
-            throw new IOException("Missing lakeFS access key");
-        }
-        String secretKey = conf.get(FS_LAKEFS_SECRET_KEY);
-        if (secretKey == null) {
-            throw new IOException("Missing lakeFS secret key");
-        }
-        this.apiClient = io.lakefs.clients.api.Configuration.getDefaultApiClient();
-        this.apiClient.setBasePath(endpoint);
-        HttpBasicAuth basicAuth = (HttpBasicAuth)this.apiClient.getAuthentication(BASIC_AUTH);
-        basicAuth.setUsername(accessKey);
-        basicAuth.setPassword(secretKey);
+	LOG.debug("$$$$$$$$$$$$$$$$$$$$$$$$$$$$ initialize: {} $$$$$$$$$$$$$$$$$$$$$$$$$$$$", name);
+
+	String host = name.getHost();
+	if (host == null) {
+	    throw new IOException("Invalid repository specified");
+	}
+	setConf(conf);
+	this.uri = name;
+
+	// setup lakeFS api client
+	String endpoint = conf.get(FS_LAKEFS_ENDPOINT, "http://localhost:8000/api/v1");
+	String accessKey = conf.get(FS_LAKEFS_ACCESS_KEY);
+	if (accessKey == null) {
+	    throw new IOException("Missing lakeFS access key");
+	}
+	String secretKey = conf.get(FS_LAKEFS_SECRET_KEY);
+	if (secretKey == null) {
+	    throw new IOException("Missing lakeFS secret key");
+	}
+	this.apiClient = io.lakefs.clients.api.Configuration.getDefaultApiClient();
+	this.apiClient.setBasePath(endpoint);
+	HttpBasicAuth basicAuth = (HttpBasicAuth)this.apiClient.getAuthentication(BASIC_AUTH);
+	basicAuth.setUsername(accessKey);
+	basicAuth.setPassword(secretKey);
+
+	s3Client = createS3ClientFromConf(conf);
     }
 
+    /**
+     * @return an Amazon S3 client configured much like S3A configure theirs.
+     */
+    static private AmazonS3 createS3ClientFromConf(Configuration conf) {
```

Comment:
A suggestion: you can create a UndelyingFSFactory class and move the s3 client creation there.